[HDInsight] Fixed descriptions, added support for managed disk clusters, added OMS operations, fixed examples, removed Windows RDP operation

wawon-msft opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

Commit:
[HDInsight] Fixed descriptions to use proper capitalization and phrasing.

Commit:
[HDInsight] Various fixes and updates: removed patching example, added OMS (extensions) operations, removed Windows RDP operation, added support for managed disks clusters, and fixed naming conventions.

wawon-msft committed:
[HDInsight] Fixed paths to examples and fixed expected response of an OMS operation.

wawon-msft committed:
[HDInsight] Removed duplicate example (CreateWindowsHadoopRdp.json).

wawon-msft committed:
[HDInsight] Replaced extends with allOf and fixed indentation.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Azure/azure-rest-api-specs?pullRequest=1749) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Azure/azure-rest-api-specs?pullRequest=1749) before we can accept your contribution.<br/><br/>:x: wawon-msft [sign now](https://cla.opensource.microsoft.com/Azure/azure-rest-api-specs?pullRequest=1749)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Azure/azure-rest-api-specs?pullRequest=1749) it.</sub>

anuchandy commented:
@wawon-msft there are 3 model validation [errors](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/280235938) reported, please fix them.

anuchandy commented on specification/hdinsight/resource-manager/Microsoft.HDInsight/2015-03-01-preview/cluster.json:
```diff
@@ -96,62 +93,12 @@
-                "tags": [
-                    "Clusters"
-                ],
-                "operationId": "Clusters_Update",
```
Why this operation is removed?

(1). Is it the case that RP never supported this operation hence having this operation in the swagger is simply a bug in the spec and you are just fixing it? OR (2). RP used to support this operation but not supported anymore?

If (2) is the case then it is a breaking change, which requires api-version upgrade. Please confirm.

wawon-msft replied:
Hmm, good catch. This must have been an error in the commit, since we still support patching.

anuchandy commented on specification/hdinsight/resource-manager/Microsoft.HDInsight/2015-03-01-preview/cluster.json:
```diff
@@ -505,6 +451,13 @@
+                        "$ref": "#/definitions/DataDisksGroups"
+                    },
+                    "description": "The data disks groups for the role."
+                },
```
Addition of a new property without upgrading api-version is a breaking change.

wawon-msft replied:
This is an optional property and for this specific case, there should be no compatibility issues with prior SDKs.

anuchandy replied:
It seems this model is used only for `Cluster_Create` operation 
which does a PUT on the endpoint `"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.HDInsight/clusters/{clusterName}"`. 

Can customer use the same method to update a cluster? i mean can user get an existing Cluster instance, modify it locally and specify modified instance as a PUT payload (via `Cluster_Create` call)?

wawon-msft replied:
A customer can't update their cluster with that same PUT call, e.g. if a customer creates a cluster with data disks, they cannot update it to remove the data disks. They would have to delete and create a separate cluster without data disks to do so.

anuchandy replied:
Fine, if that is the case then this change won't cause compatibility issues with prior SDK generated for the same api version.

anuchandy reviewed (commented):
Seems PR contains breaking changes (removal and addition of features), tagging for ARM feedback.  @ravbhatnagar please take a look.

wawon-msft committed:
[HDInsight] Re-added patching for clusters and added examples for cluster monitoring.

anuchandy commented:
@ravbhatnagar need your sign-off since in general addition of property is considered as breaking change. But in this specific case I don't see it as really breaking consumers, please see the[ discussion](https://github.com/Azure/azure-rest-api-specs/pull/1749#discussion_r141416578).

ravbhatnagar commented:
@anuchandy  - Looks good.

anuchandy commented:
@ravbhatnagar Thanks. @wawon-msft as [mentioned](https://github.com/Azure/azure-rest-api-specs/pull/1749#issuecomment-332594572) in the previous comment, can you take a look at the model validation [failures](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/280609333) and fix the examples?

wawon-msft committed:
[HDInsight] Fixed responses for configurations, extensions, and script actions.

wawon-msft committed:
[HDInsight] Minor fixes in the spec for configurations and extensions.

anuchandy commented:
@wawon-msft you can use `TrackedResource` and `ProxyResource` to represent `Cluster` and `Application` respectively.

```json
    "Resource": {
        "description": "The core properties of ARM resources",
        "properties": {
            "id": {
                "readOnly": true,
                "type": "string",
                "description": "Fully qualified resource Id for the resource."
            },
            "name": {
                "readOnly": true,
                "type": "string",
                "description": "The name of the resource"
            },
            "type": {
                "readOnly": true,
                "type": "string",
                "description": "The type of the resource."
            }
        },
        "x-ms-azure-resource": true
    },
    "TrackedResource": {
        "description": "The resource model definition for a ARM tracked top level resource",
        "properties": {
            "location": {
                "type": "string",
                "x-ms-mutability": [
                    "read",
                    "create"
                ],
                "description": "The Azure Region where the resource lives"
            },
            "tags": {
                "type": "object",
                "additionalProperties": {
                    "type": "string"
                },
                "x-ms-mutability": [
                    "read",
                    "create",
                    "update"
                ],
                "description": "Resource tags."
            }
        },
        "allOf": [
            {
                "$ref": "#/definitions/Resource"
            }
        ]
    },
    "ProxyResource": {
        "description": "The resource model definition for a ARM proxy resource. It will have everything other than required location and tags",
        "allOf": [
            {
                "$ref": "#/definitions/Resource"
            }
        ]
    }
```

Now derive Cluster and Application from it.

```json
    "Cluster": {
        "allOf": [
            {
                "$ref": "#/definitions/TrackedResource"
            }
        ],
        "properties": {
            "etag": {
                "type": "string",
                "description": "The ETag for the resource"
            },
            "properties": {
                "$ref": "#/definitions/ClusterGetProperties",
                "description": "The properties of the cluster."
            }
        },
        "description": "The HDInsight cluster."
    },

    
    "Application": {
        "description": "The HDInsight cluster application",
        "type": "object",
        "allOf": [
            {
                "$ref": "#/definitions/ProxyResource"
            }
        ],
        "properties": {
            "etag": {
                "description": "The ETag for the application",
                "type": "string"
            },
            "tags": {
                "description": "The tags for the application.",
                "type": "object",
                "additionalProperties": {
                    "type": "string"
                }
            },
            "properties": {
                "$ref": "#/definitions/ApplicationGetProperties",
                "description": "The properties of the application."
            }
        }
    }

```

This should fix this (violation)[https://travis-ci.org/Azure/azure-rest-api-specs/jobs/282927260#L664].

Also can you double check that application model really has `tags` in it? usually only the top level resource  will have tags, Application seems a child of top-level Cluster.

wawon-msft committed:
[HDInsight] Updated spec to use TrackedResource and ProxyResource for Cluster and Application.

wawon-msft commented:
Thanks for the suggestion! I've updated the pull request accordingly.

You are correct, application is not a top level resource. I've removed the tags from it.

anuchandy commented:
@wawon-msft cool. there is still some model validation, please take a look.

wawon-msft commented:
Hmm, it looks like applications actually _do_ support tags, but only for creates and gets (one cannot update their tags). I'm going to add the tags back in for now.

wawon-msft committed:
[HDInsight] Added back tags that were removed from previous commit.

anuchandy commented:
@wawon-msft there are examples [missing](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/283001887#L810) for bunch of operations (for Extension_Get, Extension_Delete, ScriptActions_Delete and ScriptExecutionHistory_Promote). Please provide them.

There are bunch of `RPCViolation` `DefinitionsPropertiesNamesCamelCase` reported by linter, since making those changes are breaking, please consider fixing them in the future RP version.

wawon-msft committed:
[HDInsight] Added missing examples.

wawon-msft committed:
[HDInsight] Fixed expected response for cluster resize, and added better descriptions for properties of SecurityProfile.

wawon-msft committed:
[HDInsight] Fixed validation issues with the resize operation from the previous commit.

wawon-msft commented:
Thanks for the heads up, I've updated the pull request. The only linter errors now are the camelcase ones.

anuchandy commented:
@wawon-msft thank you, merging PR.

AutorestCI commented:
No modification for AutorestCI/azure-sdk-for-python

AutorestCI commented:
No modification for AutorestCI/azure-sdk-for-ruby

AutorestCI commented:
No modification for AutorestCI/azure-sdk-for-node
